File: pc-server-dotnet/Network/WebSocketServer.cs
using System.Net;
using System.Net.Security;
using System.Net.WebSockets;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Text.Json;
using System.Threading;
using Microsoft.Extensions.Logging;
using AppConnectServer.Core;
using AppConnectServer.Clipboard;

namespace AppConnectServer.Network;

public class WebSocketServer
{
    private readonly ILogger<WebSocketServer> _logger;
    private readonly Configuration _config;
    private readonly X509Certificate2? _certificate;
    private readonly KeyExchangeManager _keyExchange;
    private readonly Dictionary<WebSocket, ConnectionState> _connections = new();
    private HttpListener? _listener;
    private CancellationTokenSource? _cancellationTokenSource;

    public WebSocketServer(
        X509Certificate2? certificate,
        Configuration config,
        ILoggerFactory loggerFactory,
        Action<string>? onClipboardReceived = null)
    {
        _certificate = certificate;
        _config = config;
        _logger = loggerFactory.CreateLogger<WebSocketServer>();
        OnClipboardReceived = onClipboardReceived;
        _keyExchange = new KeyExchangeManager(_config.RsaKeyFile, loggerFactory);
    }

    public Action<string>? OnClipboardReceived { get; set; }
    
    public void Stop()
    {
        _cancellationTokenSource?.Cancel();
        try
        {
            _listener?.Stop();
            _listener?.Close();
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Error stopping HttpListener");
        }
    }

    public async Task StartAsync(string host, int port)
    {
        _logger.LogInformation("Starting WebSocket server on {Host}:{Port}", host, port);

        _cancellationTokenSource = new CancellationTokenSource();
        _listener = new HttpListener();
        
        // Note: HttpListener with HTTPS requires certificate binding
        // For now, we'll use HTTP and add HTTPS support later
        // On macOS/Linux, HttpListener may require elevated permissions for ports <= 1024
        // For ports > 1024, it should work without sudo
        
        // Validate port
        if (port <= 0 || port > 65535)
        {
            throw new ArgumentException($"Invalid port number: {port}. Port must be between 1 and 65535.");
        }
        
        // Warn if using privileged port on Unix-like systems
        if (port <= 1024 && (Environment.OSVersion.Platform == PlatformID.Unix || 
                            Environment.OSVersion.Platform == PlatformID.MacOSX))
        {
            _logger.LogWarning(
                "Port {Port} is a privileged port. On macOS/Linux, HttpListener may require elevated permissions. " +
                "Consider using a port > 1024 to avoid permission issues.", port);
        }
        
        var prefix = $"http://{host}:{port}/";
        
        try
        {
            _listener.Prefixes.Add(prefix);
            
            // Try to start HttpListener with better error handling
            try
            {
                _listener.Start();
                _logger.LogInformation("WebSocket server running on {Prefix}", prefix);
            }
            catch (HttpListenerException ex)
            {
                _logger.LogError(ex, 
                    "Failed to start HttpListener. This may require elevated permissions on macOS/Linux. " +
                    "Error code: {ErrorCode}, Message: {Message}", ex.ErrorCode, ex.Message);
                throw new InvalidOperationException(
                    $"Cannot start HTTP server on {prefix}. " +
                    $"On macOS/Linux, you may need to run with elevated permissions or use a port > 1024. " +
                    $"Original error: {ex.Message}", ex);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Unexpected error starting HttpListener");
                throw;
            }

            // Accept connections in a loop
            // Note: GetContextAsync doesn't support cancellation tokens directly,
            // so we check cancellation between accepts
            while (!_cancellationTokenSource.Token.IsCancellationRequested)
            {
                try
                {
                    // Accept connection (this blocks until a connection arrives)
                    var context = await _listener.GetContextAsync();
                    
                    // Check if we should still process this connection
                    if (_cancellationTokenSource.Token.IsCancellationRequested)
                    {
                        try
                        {
                            context.Response.StatusCode = 503; // Service Unavailable
                            context.Response.Close();
                        }
                        catch { }
                        break;
                    }
                    
                    // Handle connection in background task with proper async/await
                    _ = Task.Run(async () =>
                    {
                        try
                        {
                            await HandleContextAsync(context, _cancellationTokenSource.Token);
                        }
                        catch (Exception ex)
                        {
                            _logger.LogError(ex, "Error in HandleContextAsync task");
                        }
                    }, _cancellationTokenSource.Token);
                }
                catch (HttpListenerException ex)
                {
                    if (_cancellationTokenSource.Token.IsCancellationRequested)
                    {
                        // Expected when shutting down
                        _logger.LogDebug("HttpListener stopped (shutdown requested)");
                        break;
                    }
                    _logger.LogWarning(ex, "HttpListener error: {Message} (ErrorCode: {ErrorCode})", 
                        ex.Message, ex.ErrorCode);
                    // Continue to accept more connections
                }
                catch (ObjectDisposedException)
                {
                    // Listener was disposed, expected during shutdown
                    _logger.LogDebug("HttpListener was disposed");
                    break;
                }
                catch (InvalidOperationException ex)
                {
                    if (_cancellationTokenSource.Token.IsCancellationRequested)
                    {
                        break;
                    }
                    _logger.LogWarning(ex, "HttpListener invalid operation: {Message}", ex.Message);
                    break;
                }
                catch (Exception ex)
                {
                    if (_cancellationTokenSource.Token.IsCancellationRequested)
                    {
                        break; // Expected when shutting down
                    }
                    _logger.LogError(ex, "Unexpected error accepting connection: {Message}", ex.Message);
                    // Wait a bit before trying again to avoid tight loop
                    try
                    {
                        await Task.Delay(1000, _cancellationTokenSource.Token);
                    }
                    catch (OperationCanceledException)
                    {
                        break;
                    }
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "WebSocket server error");
            throw;
        }
        finally
        {
            // Clean up HttpListener
            try
            {
                _listener?.Stop();
                _listener?.Close();
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Error stopping HttpListener");
            }
        }
    }

    private async Task HandleContextAsync(HttpListenerContext context, CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return;
        }

        if (!context.Request.IsWebSocketRequest)
        {
            try
            {
                context.Response.StatusCode = 400;
                context.Response.Close();
            }
            catch (Exception ex)
            {
                _logger.LogDebug(ex, "Error closing non-WebSocket request");
            }
            return;
        }

        WebSocketContext? wsContext = null;
        try
        {
            // Accept WebSocket connection with proper error handling
            wsContext = await context.AcceptWebSocketAsync(null);
            
            if (wsContext?.WebSocket != null)
            {
                await HandleClientAsync(wsContext.WebSocket, cancellationToken);
            }
        }
        catch (HttpListenerException ex)
        {
            _logger.LogWarning(ex, "HttpListener error during WebSocket acceptance: {Message}", ex.Message);
            try
            {
                context.Response.StatusCode = 500;
                context.Response.Close();
            }
            catch { }
        }
        catch (WebSocketException ex)
        {
            _logger.LogWarning(ex, "WebSocket error during connection: {Message}", ex.Message);
            try
            {
                context.Response.StatusCode = 500;
                context.Response.Close();
            }
            catch { }
        }
        catch (InvalidOperationException ex)
        {
            _logger.LogWarning(ex, "Invalid operation during WebSocket acceptance: {Message}", ex.Message);
            try
            {
                context.Response.StatusCode = 500;
                context.Response.Close();
            }
            catch { }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Unexpected error handling WebSocket connection");
            try
            {
                context.Response.StatusCode = 500;
                context.Response.Close();
            }
            catch { }
        }
        finally
        {
            try
            {
                wsContext?.WebSocket?.Dispose();
            }
            catch (Exception ex)
            {
                _logger.LogDebug(ex, "Error disposing WebSocket");
            }
        }
    }

    private async Task HandleClientAsync(WebSocket websocket, CancellationToken cancellationToken)
    {
        var remoteAddr = websocket.State == WebSocketState.Open 
            ? websocket.GetHashCode().ToString() 
            : "unknown";
        
        _logger.LogInformation("WebSocket client connected: {RemoteAddr}", remoteAddr);

        var connectionState = new ConnectionState
        {
            Address = remoteAddr,
            ConnectedAt = DateTime.UtcNow
        };
        _connections[websocket] = connectionState;

        try
        {
            // Handle key exchange
            await HandleKeyExchangeAsync(websocket, connectionState, cancellationToken);

            // Send connection status
            await SendConnectionStatusAsync(websocket, "connected");

            // Handle messages
            await HandleMessagesAsync(websocket, connectionState, cancellationToken);
        }
        catch (WebSocketException ex)
        {
            _logger.LogInformation("WebSocket client disconnected: {RemoteAddr} - {Error}", remoteAddr, ex.Message);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error handling client {RemoteAddr}", remoteAddr);
        }
        finally
        {
            _connections.Remove(websocket);
            try
            {
                await websocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", cancellationToken);
            }
            catch { }
        }
    }

    private async Task HandleKeyExchangeAsync(WebSocket websocket, ConnectionState state, CancellationToken cancellationToken)
    {
        // Check for pre-shared key (for testing)
        var preSharedKey = Environment.GetEnvironmentVariable("APPCONNECT_PRE_SHARED_KEY");
        if (!string.IsNullOrEmpty(preSharedKey))
        {
            try
            {
                var keyBytes = Convert.FromBase64String(preSharedKey);
                if (keyBytes.Length == EncryptionManager.KeySize)
                {
                    state.Encryption = new EncryptionManager();
                    state.Encryption.SetKey(keyBytes);
                    state.KeyExchanged = true;
                    _logger.LogInformation("Using pre-shared key for {Address}", state.Address);
                    return;
                }
            }
            catch { }
        }

        // Perform RSA key exchange
        try
        {
            var buffer = new byte[4096];
            var result = await websocket.ReceiveAsync(new ArraySegment<byte>(buffer), cancellationToken);
            
            if (result.MessageType == WebSocketMessageType.Close)
            {
                return;
            }

            var message = Encoding.UTF8.GetString(buffer, 0, result.Count);
            var keyExchangeData = JsonSerializer.Deserialize<JsonElement>(message);

            if (keyExchangeData.TryGetProperty("type", out var type) && 
                type.GetString() == "key_exchange" &&
                keyExchangeData.TryGetProperty("encrypted_key", out var encryptedKey))
            {
                var encryptedKeyB64 = encryptedKey.GetString();
                if (string.IsNullOrEmpty(encryptedKeyB64))
                {
                    throw new InvalidOperationException("Missing encrypted_key in key exchange");
                }

                var aesKey = _keyExchange.DecryptAesKey(encryptedKeyB64);
                
                state.Encryption = new EncryptionManager();
                state.Encryption.SetKey(aesKey);
                state.KeyExchanged = true;

                // Send confirmation
                var ack = JsonSerializer.Serialize(new { type = "key_exchange_ack", status = "ok" });
                await websocket.SendAsync(
                    Encoding.UTF8.GetBytes(ack),
                    WebSocketMessageType.Text,
                    true,
                    cancellationToken);

                _logger.LogInformation("Key exchange completed for {Address}", state.Address);
            }
            else
            {
                throw new InvalidOperationException("Expected key exchange message");
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Key exchange failed");
            try
            {
                var error = JsonSerializer.Serialize(new { type = "key_exchange_ack", status = "error", message = ex.Message });
                await websocket.SendAsync(
                    Encoding.UTF8.GetBytes(error),
                    WebSocketMessageType.Text,
                    true,
                    cancellationToken);
            }
            catch { }
            throw;
        }
    }

    private async Task HandleMessagesAsync(WebSocket websocket, ConnectionState state, CancellationToken cancellationToken)
    {
        var buffer = new byte[4096];
        
        while (websocket.State == WebSocketState.Open && !cancellationToken.IsCancellationRequested)
        {
            var result = await websocket.ReceiveAsync(new ArraySegment<byte>(buffer), cancellationToken);
            
            if (result.MessageType == WebSocketMessageType.Close)
            {
                break;
            }

            if (result.MessageType == WebSocketMessageType.Text)
            {
                var message = Encoding.UTF8.GetString(buffer, 0, result.Count);
                await ProcessMessageAsync(websocket, message, state, cancellationToken);
            }
        }
    }

    private async Task ProcessMessageAsync(WebSocket websocket, string message, ConnectionState state, CancellationToken cancellationToken)
    {
        if (!state.KeyExchanged || state.Encryption == null)
        {
            _logger.LogWarning("Received message before key exchange");
            return;
        }

        // Check if this is a control message (plain JSON) or encrypted clipboard data
        if (!message.Contains('|'))
        {
            try
            {
                var controlMsg = JsonSerializer.Deserialize<JsonElement>(message);
                if (controlMsg.TryGetProperty("type", out _))
                {
                    await HandleControlMessageAsync(websocket, controlMsg, state);
                    return;
                }
            }
            catch { }
        }

        // This is encrypted clipboard data
        try
        {
            var decrypted = state.Encryption.DecryptFromTransmission(message);
            var clipboardItem = JsonSerializer.Deserialize<ClipboardItem>(decrypted);
            
            if (clipboardItem != null && !string.IsNullOrEmpty(clipboardItem.Content))
            {
                _logger.LogInformation("Received clipboard: {Content}...", 
                    clipboardItem.Content.Length > 50 
                        ? clipboardItem.Content.Substring(0, 50) 
                        : clipboardItem.Content);

                // Write to clipboard
                ClipboardWriter.WriteText(clipboardItem.Content);

                // Notify callback
                OnClipboardReceived?.Invoke(clipboardItem.Content);

                state.MessagesReceived++;
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing message");
        }
    }

    private async Task HandleControlMessageAsync(WebSocket websocket, JsonElement message, ConnectionState state)
    {
        if (message.TryGetProperty("type", out var type))
        {
            var typeStr = type.GetString();
            _logger.LogDebug("Received control message: {Type}", typeStr);
            // Handle control messages (error_report, connection_status, etc.)
        }
    }

    public async Task<bool> SendClipboardAsync(WebSocket websocket, string content)
    {
        if (!_connections.TryGetValue(websocket, out var state) || 
            !state.KeyExchanged || 
            state.Encryption == null)
        {
            return false;
        }

        try
        {
            var clipboardItem = new ClipboardItem
            {
                Content = content,
                SourceDeviceId = _config.DeviceName,
                Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                Hash = EncryptionManager.CalculateHash(content),
                Ttl = 86400000 // 24 hours
            };

            var json = JsonSerializer.Serialize(clipboardItem);
            var encrypted = state.Encryption.EncryptForTransmission(json);

            var bytes = Encoding.UTF8.GetBytes(encrypted);
            await websocket.SendAsync(
                new ArraySegment<byte>(bytes),
                WebSocketMessageType.Text,
                true,
                CancellationToken.None);

            state.MessagesSent++;
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error sending clipboard");
            return false;
        }
    }

    public async Task<int> BroadcastClipboardAsync(string content)
    {
        var count = 0;
        var tasks = _connections.Keys
            .Where(ws => ws.State == WebSocketState.Open)
            .Select(async ws =>
            {
                if (await SendClipboardAsync(ws, content))
                {
                    Interlocked.Increment(ref count);
                }
            });

        await Task.WhenAll(tasks);
        return count;
    }

    private async Task SendConnectionStatusAsync(WebSocket websocket, string status)
    {
        if (!_connections.TryGetValue(websocket, out var state))
        {
            return;
        }

        try
        {
            var statusMsg = JsonSerializer.Serialize(new
            {
                type = "connection_status",
                status = status,
                timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            });

            var bytes = Encoding.UTF8.GetBytes(statusMsg);
            await websocket.SendAsync(
                new ArraySegment<byte>(bytes),
                WebSocketMessageType.Text,
                true,
                CancellationToken.None);
        }
        catch { }
    }

    private class ConnectionState
    {
        public string Address { get; set; } = string.Empty;
        public EncryptionManager? Encryption { get; set; }
        public bool KeyExchanged { get; set; }
        public DateTime ConnectedAt { get; set; }
        public int MessagesSent { get; set; }
        public int MessagesReceived { get; set; }
    }
}

public class ClipboardItem
{
    public string Content { get; set; } = string.Empty;
    public string SourceDeviceId { get; set; } = string.Empty;
    public long Timestamp { get; set; }
    public string Hash { get; set; } = string.Empty;
    public long Ttl { get; set; }
}

File: pc-server/src/websocket_server.py
"""WebSocket server with key exchange and message handling."""
import asyncio
import json
import base64
import logging
from pathlib import Path
from typing import Optional, Dict, Callable
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives import hashes
import websockets
from websockets.asyncio.server import ServerConnection
from websockets.exceptions import ConnectionClosed

from .encryption import EncryptionManager
from .message_parser import ClipboardItem
from .clipboard_writer import ClipboardWriter

logger = logging.getLogger(__name__)


class WebSocketServer:
    """Secure WebSocket server for clipboard synchronization."""
    
    def __init__(self, ssl_context, rsa_private_key_file: Path, 
                 device_id: str, on_clipboard_received: Optional[Callable[[str], None]] = None,
                 pre_shared_key: Optional[bytes] = None):
        """
        Initialize WebSocket server.
        
        Args:
            ssl_context: SSL context for WSS
            rsa_private_key_file: Path to RSA private key for key exchange
            device_id: Device identifier for sourceDeviceId
            on_clipboard_received: Optional callback when clipboard is received from Android
            pre_shared_key: Optional pre-shared AES key (for testing without key exchange)
        """
        self.ssl_context = ssl_context
        self.rsa_private_key_file = rsa_private_key_file
        self.device_id = device_id
        self.on_clipboard_received = on_clipboard_received
        self.pre_shared_key = pre_shared_key
        
        # Per-connection state
        self.connections: Dict[ServerConnection, Dict] = {}
        
        # Load RSA private key (only if not using pre-shared key)
        self._rsa_private_key = None
        if not pre_shared_key:
            self._load_rsa_key()
        
    def _load_rsa_key(self) -> None:
        """Load RSA private key for key exchange."""
        if not self.rsa_private_key_file.exists():
            raise FileNotFoundError(
                f"RSA private key file not found: {self.rsa_private_key_file}. "
                f"Please ensure the key file exists or run the server to generate it."
            )
        
        try:
            with open(self.rsa_private_key_file, "rb") as f:
                key_data = f.read()
                if not key_data:
                    raise ValueError(f"RSA private key file is empty: {self.rsa_private_key_file}")
                
                self._rsa_private_key = serialization.load_pem_private_key(
                    key_data,
                    password=None
                )
                logger.info(f"Loaded RSA private key from {self.rsa_private_key_file}")
        except FileNotFoundError:
            raise
        except Exception as e:
            logger.error(f"Failed to load RSA private key from {self.rsa_private_key_file}: {e}")
            raise ValueError(f"Invalid RSA private key file: {e}") from e
            
    async def handle_client(self, websocket: ServerConnection) -> None:
        """Handle a WebSocket client connection."""
        remote_addr = websocket.remote_address
        logger.info(f"WebSocket client connected: {remote_addr}")
        
        # Initialize connection state
        connection_state = {
            "encryption": None,  # EncryptionManager instance
            "key_exchanged": False,
            "address": remote_addr
        }
        self.connections[websocket] = connection_state
        
        try:
            # Wait for key exchange message
            await self._handle_key_exchange(websocket, connection_state)
            
            # Handle messages
            async for message in websocket:
                await self._handle_message(websocket, message, connection_state)
                
        except ConnectionClosed:
            logger.info(f"WebSocket client disconnected: {remote_addr}")
        except Exception as e:
            logger.error(f"Error handling client {remote_addr}: {e}", exc_info=True)
        finally:
            # Cleanup
            if websocket in self.connections:
                del self.connections[websocket]
                
    async def _handle_key_exchange(self, websocket: ServerConnection, 
                                   connection_state: Dict) -> None:
        """Handle RSA-based key exchange or use pre-shared key."""
        # If pre-shared key is set, use it directly
        if self.pre_shared_key:
            encryption = EncryptionManager(self.pre_shared_key)
            connection_state["encryption"] = encryption
            connection_state["key_exchanged"] = True
            logger.info(f"Using pre-shared key for {connection_state['address']}")
            return
            
        # Otherwise, perform RSA key exchange
        try:
            # Wait for encrypted AES key from client
            message = await asyncio.wait_for(websocket.recv(), timeout=10.0)
            
            # Parse key exchange message
            # Format: {"type": "key_exchange", "encrypted_key": "base64_encrypted_aes_key"}
            key_exchange_data = json.loads(message)
            
            if key_exchange_data.get("type") != "key_exchange":
                raise ValueError("Expected key exchange message")
                
            encrypted_key_b64 = key_exchange_data.get("encrypted_key")
            if not encrypted_key_b64:
                raise ValueError("Missing encrypted_key in key exchange")
            
            # Add padding for Base64 decode (Android uses NO_WRAP which strips padding)
            def add_padding(s: str) -> str:
                missing_padding = len(s) % 4
                if missing_padding:
                    return s + '=' * (4 - missing_padding)
                return s
                
            # Decode and decrypt AES key
            logger.debug(f"Received encrypted key (length: {len(encrypted_key_b64)})")
            encrypted_key = base64.b64decode(add_padding(encrypted_key_b64))
            logger.debug(f"Decoded encrypted key (length: {len(encrypted_key)} bytes)")
            aes_key = self._rsa_private_key.decrypt(
                encrypted_key,
                padding.OAEP(
                    mgf=padding.MGF1(algorithm=hashes.SHA256()),
                    algorithm=hashes.SHA256(),
                    label=None
                )
            )
            
            # Validate decrypted AES key size (must be 32 bytes for AES-256)
            if len(aes_key) != 32:
                raise ValueError(f"Invalid AES key size: {len(aes_key)} bytes (expected 32)")
            
            # Initialize encryption manager with shared key
            encryption = EncryptionManager(aes_key)
            connection_state["encryption"] = encryption
            connection_state["key_exchanged"] = True
            
            # Send confirmation
            await websocket.send(json.dumps({"type": "key_exchange_ack", "status": "ok"}))
            
            logger.info(f"Key exchange completed for {connection_state['address']}")
            
        except asyncio.TimeoutError:
            logger.warning("Key exchange timeout")
            raise
        except Exception as e:
            logger.error(f"Key exchange failed: {e}", exc_info=True)
            # Try to send error response to client before closing connection
            try:
                await websocket.send(json.dumps({
                    "type": "key_exchange_ack", 
                    "status": "error", 
                    "message": str(e)
                }))
            except Exception:
                # Connection may already be closed, ignore
                pass
            # Re-raise to close the connection (this is expected behavior on error)
            raise
            
    async def _handle_message(self, websocket: ServerConnection, 
                             message: str, connection_state: Dict) -> None:
        """Handle incoming encrypted message."""
        if not connection_state["key_exchanged"]:
            logger.warning("Received message before key exchange")
            return
            
        try:
            encryption = connection_state["encryption"]
            
            # Decrypt message
            decrypted_json = encryption.decrypt_from_transmission(message)
            
            # Parse ClipboardItem
            clipboard_item = ClipboardItem.from_json(decrypted_json)
            
            logger.info(f"Received clipboard: {clipboard_item.content[:50]}...")
            
            # Write to clipboard in executor to avoid blocking the event loop
            # This prevents clipboard operations from causing connection timeouts
            loop = asyncio.get_event_loop()
            try:
                await loop.run_in_executor(
                    None,
                    ClipboardWriter.write_text,
                    clipboard_item.content
                )
                logger.debug("Successfully wrote to clipboard")
            except Exception as e:
                logger.error(f"Error writing to clipboard: {e}", exc_info=True)
                # Continue even if clipboard write fails
            
            # Notify callback if set (callback is synchronous, run in executor to avoid blocking)
            if self.on_clipboard_received:
                try:
                    # Run callback in executor to avoid blocking the event loop
                    await loop.run_in_executor(
                        None,
                        lambda: self.on_clipboard_received(clipboard_item.content)
                    )
                except Exception as e:
                    logger.error(f"Error in clipboard received callback: {e}", exc_info=True)
                    # Continue even if callback fails
                
        except Exception as e:
            logger.error(f"Error handling message: {e}", exc_info=True)
            # Don't re-raise - we want to continue handling messages
            
    async def send_clipboard(self, websocket: ServerConnection, 
                            content: str) -> bool:
        """
        Send clipboard content to connected client.
        
        Args:
            websocket: WebSocket connection
            content: Clipboard text content
            
        Returns:
            True if sent successfully, False otherwise
        """
        if websocket not in self.connections:
            return False
            
        connection_state = self.connections[websocket]
        
        if not connection_state["key_exchanged"]:
            logger.warning("Cannot send: key exchange not completed")
            return False
            
        try:
            encryption = connection_state["encryption"]
            
            # Create ClipboardItem
            clipboard_item = ClipboardItem.create_text_item(
                content=content,
                source_device_id=self.device_id
            )
            
            # Encrypt and send
            encrypted_message = encryption.encrypt_for_transmission(clipboard_item.to_json())
            await websocket.send(encrypted_message)
            
            logger.info(f"Sent clipboard: {content[:50]}...")
            return True
            
        except Exception as e:
            logger.error(f"Error sending clipboard: {e}", exc_info=True)
            return False
            
    async def broadcast_clipboard(self, content: str) -> int:
        """
        Broadcast clipboard content to all connected clients.
        
        Returns:
            Number of clients that received the message
        """
        count = 0
        disconnected = []
        
        for websocket in list(self.connections.keys()):
            try:
                if await self.send_clipboard(websocket, content):
                    count += 1
            except Exception:
                disconnected.append(websocket)
                
        # Clean up disconnected clients
        for ws in disconnected:
            if ws in self.connections:
                del self.connections[ws]
                
        return count
        
    def get_connection_count(self) -> int:
        """Get number of connected clients."""
        return len(self.connections)
        
    async def start(self, host: str = "0.0.0.0", port: int = 8765) -> None:
        """Start WebSocket server."""
        logger.info(f"Starting WebSocket server on {host}:{port}")
        
        async with websockets.serve(
            self.handle_client,
            host,
            port,
            ssl=self.ssl_context
        ):
            logger.info(f"WebSocket server running on wss://{host}:{port}")
            # Keep server running
            await asyncio.Future()  # Run forever

